src/version.rs

```diff
@@ -0,0 +1,147 @@
+use std::convert::TryFrom;
+
+/// Represents the major, minor, and patch (if any) versions of this interpreter.
+///
+/// This struct is usually created with [`Python::version`].
+///
+/// # Examples
+///
+
+/// ```rust
+/// # use pyo3::Python;
+/// Python::with_gil(|py| {
+///     // PyO3 supports Python 3.7 and up.
+///     assert!(py.version_info() >= (3, 7));
+///     assert!(py.version_info() >= (3, 7, 0));
+/// });
+/// ```
+///
+/// [`Python::version`]: crate::marker::Python::version
+#[derive(Debug)]
+pub struct PythonVersionInfo<'py> {
+    /// Python major version (e.g. `3`).
+    pub major: u8,
+    /// Python minor version (e.g. `11`).
+    pub minor: u8,
+    /// Python patch version (e.g. `0`).
+    pub patch: u8,
+    /// Python version suffix, if applicable (e.g. `a0`).
+    pub suffix: Option<&'py str>,
+}
+
+impl<'py> TryFrom<&'py str> for PythonVersionInfo<'py> {
```

Comment:
Note that implementing this trait rather than private `from_str` function has made this public. I think we should consider implementing the `FromStr` trait rather than `TryFrom`?

Reply:
Unfortunately we can't do that...
 > FromStr does not have a lifetime parameter, and so you can only parse types that do not contain a lifetime parameter themselves.

Reply:
Right, so if we change to `&'static` then this works. That's probably why I didn't implement `FromStr` originally.

Reply:
As we're not _entirely_ sure what the correct implementation for this type should be w.r.t. static etc, would you be ok if we rolled back this trait implementation for the moment to a private function? We can always commit again to the trait later.